src/redisearch_rs/low_memory_thin_vec/src/lib.rs

```diff
@@ -0,0 +1,2050 @@
+    /// ```
+    /// use low_memory_thin_vec::{LowMemoryThinVec, low_memory_thin_vec};
+    /// use std::convert::TryInto;
+    ///
+    /// let r: Result<[i32; 4], _> = (0..10).collect::<LowMemoryThinVec<_>>().try_into();
+    /// assert_eq!(r, Err(low_memory_thin_vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]));
+    /// ```
+    ///
+    /// If you're fine with just getting a prefix of the `LowMemoryThinVec<T>`,
+    /// you can call [`.truncate(N)`](LowMemoryThinVec::truncate) first.
+    /// ```
+    /// use low_memory_thin_vec::{LowMemoryThinVec, low_memory_thin_vec};
+    /// use std::convert::TryInto;
+    ///
+    /// let mut v = LowMemoryThinVec::from("hello world");
+    /// v.sort();
+    /// v.truncate(2);
+    /// let [a, b]: [_; 2] = v.try_into().unwrap();
+    /// assert_eq!(a, b' ');
+    /// assert_eq!(b, b'd');
+    /// ```
+    fn try_from(mut vec: LowMemoryThinVec<T>) -> Result<[T; N], LowMemoryThinVec<T>> {
+        if vec.len() != N {
+            return Err(vec);
+        }
+
+        // SAFETY: `.set_len(0)` is always sound.
+        unsafe { vec.set_len(0) };
+
+        // SAFETY: A `LowMemoryThinVec`'s pointer is always aligned properly, and
+        // the alignment the array needs is the same as the items.
+        // We checked earlier that we have sufficient items.
+        // The items will not double-drop as the `set_len`
+        // tells the `LowMemoryThinVec` not to also drop them.
+        let array = unsafe { ptr::read(vec.data_raw() as *const [T; N]) };
+        Ok(array)
+    }
+}
+
+/// An iterator that moves out of a vector.
+///
+/// This `struct` is created by the [`LowMemoryThinVec::into_iter`][]
+/// (provided by the [`IntoIterator`] trait).
+///
+/// # Example
+///
+/// ```
+/// use low_memory_thin_vec::low_memory_thin_vec;
+///
+/// let v = low_memory_thin_vec![0, 1, 2];
+/// let iter: low_memory_thin_vec::IntoIter<_> = v.into_iter();
+/// ```
+pub struct IntoIter<T> {
+    vec: LowMemoryThinVec<T>,
+    start: usize,
+}
+
+impl<T> IntoIter<T> {
+    /// Returns the remaining items of this iterator as a slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use low_memory_thin_vec::low_memory_thin_vec;
+    ///
+    /// let vec = low_memory_thin_vec!['a', 'b', 'c'];
+    /// let mut into_iter = vec.into_iter();
+    /// assert_eq!(into_iter.as_slice(), &['a', 'b', 'c']);
+    /// let _ = into_iter.next().unwrap();
+    /// assert_eq!(into_iter.as_slice(), &['b', 'c']);
+    /// ```
+    pub fn as_slice(&self) -> &[T] {
+        // SAFETY:
+        // - The data pointer is always aligned and it's safe to offset by an
+        //   index within the bounds of the vector, since it'll fall within the same
+        //   allocation.
+        //   `self.start` has been checked to be within bounds when the iterator was created/last advanced.
+        let start_ptr = unsafe { self.vec.data_raw().add(self.start) };
+        let n_remaining_elements = self.len();
```

Comment:
where does `self.len()` is defined? is it `self.vec.len() - self.start`?

Reply:
It comes from [`ExactSizeIterator`](https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html), a trait that we implement just below.
And yes, it is the number of remaining elements, as defined in `size_hint`.